noise_corrector.py

```diff
@@ -1,70 +1,177 @@
+    target = N
+    c = 0
+    while math.comb(c, k) <= N:
+      c += 1
+
+    combos = []
+    while c >= 0 and k >= 0:
+      coef = math.comb(c, k)
+      if coef <= target:
+        combos.append(c)
+        k -= 1
+        target -= coef
+      c -= 1
+    return combos
+
+  def bars_to_report(star_index, num_star):
+    """Returns a report description"""
+
+    bars_prior_to_star = star_index - (max_reports - 1 - num_star)
+
+    # Any stars before any bars means the report is suppressed.
+    if bars_prior_to_star == 0:
+      return "No report"
+
+    # Otherwise, the number of bars just indexes into the cartesian product
+    # of |data_cardinality| x |num_windows|.
+    # Subtract the first bar because that just begins the first window.
+    window, data = divmod(bars_prior_to_star - 1, data_cardinality)
+    return {"window": window, "data": data}
+
+  def generate_output(N):
+    star_indices = find_k_comb(N, max_reports)
+    return [bars_to_report(s, i) for i, s in enumerate(star_indices)]
+
+  star_bar_length = data_cardinality * max_windows + max_reports
+  num_sequences = math.comb(star_bar_length, max_reports)
+  return [generate_output(i) for i in range(num_sequences)]
+
+
+def get_possible_event_source_outputs():
+  """Returns the possible outputs for an event source"""
+  return generate_possible_outputs_per_source(1, 2, 1)
+
+
+def get_possible_nav_source_outputs():
+  """Returns all the possible outputs for navigation source"""
+  return generate_possible_outputs_per_source(3, 8, 3)
+
+
+# Params are taken from the EVENT.md explainer:
+# https://github.com/WICG/conversion-measurement-api/blob/main/EVENT.md#data-limits-and-noise
+def correct_event_sources(observed_values):
+  # No triggers, 1 trigger with metadata 0, 1 trigger with metadata 1.
+  assert len(observed_values) == 3
+  randomized_response_rate = .0000025
+  return estimate_true_values(observed_values, randomized_response_rate)
+
+
+def correct_nav_sources(observed_values):
+  assert len(observed_values) == 2925
+  randomized_response_rate = .0024
+  return estimate_true_values(observed_values, randomized_response_rate)
+
+
+def simulate_randomization(true_reports, randomized_response_rate):
+  """Simulates randomized response on true_reports"""
+  n = sum(true_reports)
+  k = len(true_reports)
+  noisy_reports = [0] * k
+
+  # This method could be implemented by doing randomized response on every
+  # individual true report. However, we can optimize this by noticing that
+  # every bucket's final result will be distributed according to a Binomial
+  # distribution with parameter n and q = (1 - x)*(v' / n) + xr
+  # where x = randomized_response_rate. Note that this is just an estimate,
+  # since the true randomized response will not be independent for each value.
+  x = randomized_response_rate
+
+  def gen_noisy(true_count):
+    q = (1 - x) * (true_count / n) + x / k
+    return numpy.random.binomial(n, q)
+  return [gen_noisy(c) for c in true_reports]
```

Comment:
Here is a possible two-phrase algorithm for this:
- first, compute the number of non-flipped reports for each possible input i in k; this is simply Binomial[true_count[i], 1 - x]
- now, we have the number of flipped reports. These we just randomly put them back into the buckets.

For your convenience, here's a possible implementation:

```
# Compute the non-flipped counts.
non_flipped_counts = [numpy.random.binomial(true_count, 1 - x) for true_count in true_reports]

# Distribute the flipped reports uniformly among all k buckets.
num_flipped = n - sum(non_flipped_counts)
flipped_counts = numpy.random.multinomial(num_flipped, [1/k] * k)

return non_flipped_counts + flipped_counts
```

Reply:
Done, thanks for the suggestion!